crates/rome_formatter/src/format_element.rs

```
@@ -1374,6 +1484,50 @@ impl PrintMode {
     }
 }
 
+/// Provides the printer with different representations for the same element so that the printer
+/// can pick the best fitting variant.
+///
+/// Best fitting is defined as the variant that takes the most horizontal space but fits on the line.
+#[derive(Clone, Eq, PartialEq)]
+pub struct BestFitting {
+    /// The different variants for this element.
+    /// The first element is the one that takes up the most space horizontally (the most flat),
+    /// The last element takes up the least space horizontally (but most horizontal space).
+    variants: Box<[FormatElement]>,
+}
+
+impl BestFitting {
+    pub fn new(variants: Vec<FormatElement>) -> Self {
+        assert!(variants.len() > 1, "The variants collection must contain at least two variants where the first is the least expanded state and the last element the most expanded option.");
```

Is it necessary to panic when someone tries to construct an invalid `BestFitting` rather than returning a result?
Would it be reasonable to expand `FormatError` to account for errors during IR construction?

If the intended use of `best_fitting` is that it's always passed a `vec!` literal rather than a dynamically built up vector, maybe `best_fitting` could be a macro that ensures there are always at least two variants at compile-time?

---

All usages in prettier are indeed using a fixed set of variants. It's unfortunate that rust doesn't allow for var args without using macros.

---

I introduced a `best_fitting` macro that checks the invariant and marked the `from_slice_unchecked` as unsafe.